src/sentry/incidents/logic.py

```diff
@@ -625,6 +627,27 @@ def update_alert_rule(
     return alert_rule
 
 
+def delete_alert_rule(alert_rule):
+    """
+    Marks an alert rule as deleted and fires off a task to actually delete it.
+    :param alert_rule:
+    """
+    if alert_rule.status in (
+        AlertRuleStatus.PENDING_DELETION.value,
+        AlertRuleStatus.DELETION_IN_PROGRESS.value,
+    ):
+        raise AlreadyDeletedError()
+
+    alert_rule.update(
+        # Randomize the name here so that we don't get unique constraint issues
+        # while waiting for the deletion to process
+        name=uuid4().get_hex(),
+        status=AlertRuleStatus.PENDING_DELETION.value,
+    )
+    tasks.delete_alert_rule.apply_async(kwargs={'alert_rule_id': alert_rule.id})
+    delete_snuba_subscription(alert_rule.subscription_id)
```

Comment:
Should the subscription be deleted in the background worker too?

Reply:
I don't have a strong opinion, but the main reason I'm using the task at all is just because I want to make sure the api doesn't take a long time to respond if an alert rule has a large number of related incidents. Deleting the subscription here doesn't take much time, so I left it out of the task.

Reply:
Are we going to tell the user that deleting a rule will delete all related incidents? If incidents are intended to be a historical artifact, could their deletion come as a surprise?

Reply:
We're not going to delete the incidents. I've set the relation to `on_delete=models.SET_NULL`, which just nulls the `alert_rule_id` value on any related incidents if the alert rule is deleted. So we just lose the reason the incident was created, which sara has confirmed is expected.